cpp/data_lang.cc

```diff
@@ -160,6 +160,23 @@ BigStr* ShellEncodeString(BigStr* s, int ysh_fallback) {
   return buf->getvalue();
 }
 
+Tuple2<int, int> Utf8DecodeOne(BigStr* s, int start) {
+  assert(0 <= start && start <= len(s));
```

Comment:
I think this should be `start < len(s)`, because `s[len(s)]` is not a valid part of the string

I think of it like a "half open interval" in math `[0, n)` is written

`0 <= i && i < n`

in both Python and C++

Reply:
Ah, logically this is true, but remember that we have a nul-terminator. So, `s[len(s)]` -- and in turn `fastfunc.Utf8DecodeOne(s, len(s))` -- is actually well defined.

I also have tests over this behaviour:

```python
s = 'h\xE2\xA1\x80\xC5\x81'

# UTF8_ERR_END_OF_STREAM = 6
self.assertEqual((-6, 0), fastfunc.Utf8DecodeOne(s, 6))
```
(https://github.com/oilshell/oil/pull/1965/files#diff-6d07d51003e574f4cbdd6d2a6ff6a13c1e60dd0c854b23bf9da960497ebef712R52) (+ an identical one for cpp)

I don't believe we rely on this from the python/mycpp side, but we do rely on it in `data_lang/j8.h`. While we could restrict `fastfunc.Utf8DecodeOne` so that $0 \le i \lt \text{len}(s)$, I don't want to have the interface deviate from `utf8_decode` for consistency.

Reply:
Hm I think it's the difference between a Python string/mycpp string (has a length), and a C string (has no length)

But if you think it's OK, I would add a comment by both asserts that says this case is well defined, just because it "looks weird"

I agree the way we use it in j8.h is a little weird, but it is correct, and is probably the best solution.